include/wlr/types/wlr_seat.h

```diff
@@ -42,12 +42,14 @@ struct wlr_seat_keyboard_grab;
 
 struct wlr_keyboard_grab_interface {
 	void (*enter)(struct wlr_seat_keyboard_grab *grab,
-			struct wlr_surface *surface);
+		struct wlr_surface *surface, uint32_t *keycodes, size_t keycodes_len,
+		uint32_t mods_depressed, uint32_t mods_latched, uint32_t mods_locked,
+		uint32_t group);
```

Comment:
You shouldn't need these arguments here. They should just be able to read the modifiers off the keyboard from the seat.

Reply:
>Should we read the current keyboard state like this? The compositor might want to filter some keys, and we might end up in an inconsistent state if the compositor is simulating keypresses. Would it be better to take the state as a parameter to wlr_seat_keyboard_enter?

Reply:
If we read modifiers from the `wlr_keyboard` here, then `wlr_seat_notify_modifiers` should also read modifiers from the `wlr_keyboard`.

Reply:
I don't think the overhead of having rootston reading modifier state anytime it wants to set keyboard focus is worth it. I think we can add that feature in with a special interface once we have a concrete use case for it. It's difficult to design it in the abstract when we will probably find a better solution later.

Reply:
So I'll change both functions and remove modifier arguments.

What about the keyboard keys? Do we want the seat to read them directly from the keyboard, like modifiers?

Reply:
Yeah I think it's better to have a single source of truth on the keyboard for keyboard state for now and evaluate any other use case separately.

Reply:
Compositors can also implement a custom keyboard interface if they want to simulate keypresses.